osu.Game/Skinning/Components/BeatmapAttributeText.cs

```diff
@@ -78,27 +103,106 @@ protected override void LoadComplete()
             }, true);
         }
 
-        private void updateBeatmapContent(WorkingBeatmap workingBeatmap)
+        private void updateAllInfo()
         {
-            valueDictionary[BeatmapAttribute.Title] = workingBeatmap.BeatmapInfo.Metadata.Title;
-            valueDictionary[BeatmapAttribute.Artist] = workingBeatmap.BeatmapInfo.Metadata.Artist;
-            valueDictionary[BeatmapAttribute.DifficultyName] = workingBeatmap.BeatmapInfo.DifficultyName;
-            valueDictionary[BeatmapAttribute.Creator] = workingBeatmap.BeatmapInfo.Metadata.Author.Username;
-            valueDictionary[BeatmapAttribute.Length] = TimeSpan.FromMilliseconds(workingBeatmap.BeatmapInfo.Length).ToFormattedDuration();
-            valueDictionary[BeatmapAttribute.RankedStatus] = workingBeatmap.BeatmapInfo.Status.GetLocalisableDescription();
-            valueDictionary[BeatmapAttribute.BPM] = workingBeatmap.BeatmapInfo.BPM.ToLocalisableString(@"F2");
-            valueDictionary[BeatmapAttribute.CircleSize] = ((double)workingBeatmap.BeatmapInfo.Difficulty.CircleSize).ToLocalisableString(@"F2");
-            valueDictionary[BeatmapAttribute.HPDrain] = ((double)workingBeatmap.BeatmapInfo.Difficulty.DrainRate).ToLocalisableString(@"F2");
-            valueDictionary[BeatmapAttribute.Accuracy] = ((double)workingBeatmap.BeatmapInfo.Difficulty.OverallDifficulty).ToLocalisableString(@"F2");
-            valueDictionary[BeatmapAttribute.ApproachRate] = ((double)workingBeatmap.BeatmapInfo.Difficulty.ApproachRate).ToLocalisableString(@"F2");
-            valueDictionary[BeatmapAttribute.StarRating] = workingBeatmap.BeatmapInfo.StarRating.ToLocalisableString(@"F2");
+            updateStarRating();
+            updateBpmAndLength();
+            updateBeatmapContent();
+            updateLabel();
+        }
+
+        private void updateDifficulty()
+        {
+            difficulty = new BeatmapDifficulty(workingBeatmap.Value.BeatmapInfo.Difficulty);
+
+            if (!DifficultyAdjust.Value) return;
+
+            foreach (var mod in mods.Value.OfType<IApplicableToDifficulty>())
+                mod.ApplyToDifficulty(difficulty);
+        }
+
+        private void updateStarRating()
+        {
+            starDifficultyCancellationSource?.Cancel();
+            starDifficultyCancellationSource = new CancellationTokenSource();
+
+            var starDifficultyTask = DifficultyAdjust.Value
+                ? difficultyCache.GetDifficultyAsync(workingBeatmap.Value.BeatmapInfo, ruleset.Value, mods.Value, starDifficultyCancellationSource.Token)
+                : difficultyCache.GetDifficultyAsync(workingBeatmap.Value.BeatmapInfo, ruleset.Value, null, starDifficultyCancellationSource.Token);
+            starDifficultyTask.ContinueWith(_ => Schedule(() =>
+            {
+                starRating = starDifficultyTask.GetResultSafely();
+                valueDictionary[BeatmapAttribute.StarRating] = (starRating?.Stars ?? -1).ToLocalisableString(@"F2");
```

Comment:
Why is -1 used as the fallback value? This will be printed to the user. May as well use something like `"(not available)"` or anything.